Add Option for Delegation to Any User

APTy opened the pull request:
This change resolves https://github.com/cloudflare/redoctober/issues/119 by adding an explicit option (`"AnyUser"`) for a delegator to allow _any user_ to decrypt requested data. This improves on the current design, where a delegator must specify an empty list of users for the same effect.
#### Adding the AnyUser Option

Commit 4093134d91a3c9764757ae6b17b634c65209fd0a adds the initial enhancement by extending the API for `Cache.AddKeyFromRecord()`. At this point, the function signature for this function is getting pretty heavy, but as it's only used once in `core/core.go`, this is probably OK (The test dependencies may be another matter). We document that delegating to an empty list of users is deprecated functionality, but still support it. Also adds CLI support for the same flag.

Edit: The function signature for `Cache.AddKeyFromRecord()` now accepts a `Usage` object to simplify it's API.
#### Enforcing the AnyUser Option

Commit 5549d4c6f6b1ca9e65204db0e95dee0e6b018308 adds enforcement (thereby breaking backwards-compatibility) of the flag by returning an error in `core/core.go` if we don't receive a list of `Users` or if the `AnyUser` flag is not set. This break the old, undesired behavior, and have to update quite a few test vectors to reflect the new API.

Thoughts on rolling these changes out incrementally to allow applications time to upgrade to the modified `/delegate` API?

Commit:
Add delegation option for any user to be able to decrypt

core: Add AnyUser field to DelegateRequest and pass to cache calls

keycache: Add AnyUser parameter to AddKeyFromRecord function signature

keycache_test: Add tests for AnyUser and update AddKeyFromRecord calls

cryptor: Update tests to AddKeyFromRecord to reflect API update

cmd/ro: Add bool flag for anyUser parameter

Note: This commit was further amended to change the AddKeyFromRecord
API to accept keycache.Usage as a parameter, which will make updating
usage simpler in the future.

Commit:
Remove backwards-compatibility for delegating without a list of users

keycache: Remove deprecated behavior in Usage.matches()

core: Add check during delegation for a User list or the AnyUser flag

tests: Add parameters to conform to new delegation requirements

kisom commented on core/core_test.go:
```diff
@@ -50,7 +50,7 @@ func TestCreate(t *testing.T) {
 func TestSummary(t *testing.T) {
 	createJson := []byte("{\"Name\":\"Alice\",\"Password\":\"Hello\"}")
-	delegateJson := []byte("{\"Name\":\"Bob\",\"Password\":\"Rob\",\"Time\":\"2h\",\"Uses\":1}")
+	delegateJson := []byte("{\"Name\":\"Bob\",\"Password\":\"Rob\",\"Time\":\"2h\",\"Uses\":1,\"AnyUser\":true}")
```
It would be useful to test a lot of these with `AnyUser: false` to ensure that it fails in that case.

APTy replied:
Good point. I added `AnyUser: true` to preserve the original behavior of the tests, but I agree that we should test the alternate path as well.

Commit:
Add tests in core for success and failure of the AnyUser parameter

core: Add unit tests scenarios

core: Check for a non-nil, but empty, Users list

keycache: Add unit tests

APTy commented:
@kisom please see my test update - I think this logic is best served in its own test, where it won't interfere with the intended behaviors of other tests.

kisom commented:
@APTy I won't have time this week to review this, but I will be able to on Monday.

jkroll-cf commented:
I'm aware of one project abusing the current behavior (no users specified => all users have access). I'll update once they've fixed. Meanwhile, please don't merge this.

APTy committed:
Move enforcement of Users/AnyUser flag to keycache, pass through error

APTy committed:
Use keyed fields in composite literal, per `vet` source report

Output before change:
cryptor/cryptor_test.go:112:
github.com/cloudflare/redoctober/keycache.Usage composite literal uses
unkeyed fields

kisom commented:
This has merge conflicts and needs to be rebased off master.

ryanskidmore commented:
This project is archived and will no longer receive any updates.
